token/src/components/token/erc721/erc721_approval.cairo

```diff
@@ -0,0 +1,251 @@
+    }
+
+
+    #[embeddable_as(ERC721ApprovalImpl)]
+    impl ERC721Approval<
+        TContractState,
+        +HasComponent<TContractState>,
+        +IWorldProvider<TContractState>,
+        impl ERC721Owner: erc721_owner_comp::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of IERC721Approval<ComponentState<TContractState>> {
+        fn get_approved(
+            ref self: ComponentState<TContractState>, token_id: u128
+        ) -> ContractAddress {
+            self._get_approved(token_id).address
+        }
+
+        fn is_approved_for_all(
+            self: @ComponentState<TContractState>, owner: ContractAddress, operator: ContractAddress
+        ) -> bool {
+            self.get_approval_for_all(owner, operator).approved
+        }
+
+        fn approve(ref self: ComponentState<TContractState>, to: ContractAddress, token_id: u128) {
+            let mut erc721_owner = get_dep_component_mut!(ref self, ERC721Owner);
+
+            let owner = erc721_owner.get_owner(token_id).address;
+            let caller = get_caller_address();
+            assert(
+                owner == caller || self.get_approval_for_all(owner, caller).approved,
+                Errors::UNAUTHORIZED
+            );
+            self.set_token_approval(owner, to, token_id, true)
+        }
+
+        fn set_approval_for_all(
+            ref self: ComponentState<TContractState>, operator: ContractAddress, approved: bool
+        ) {
+            self._set_approval_for_all(get_caller_address(), operator, approved)
+        }
+    }
+
+    #[embeddable_as(ERC721ApprovalCamelImpl)]
+    impl ERC721ApprovalCamel<
+        TContractState,
+        +HasComponent<TContractState>,
+        +IWorldProvider<TContractState>,
+        impl ERC721Owner: erc721_owner_comp::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of IERC721ApprovalCamel<ComponentState<TContractState>> {
+        fn getApproved(ref self: ComponentState<TContractState>, tokenId: u128) -> ContractAddress {
+            self.get_approved(tokenId)
+        }
+
+        fn isApprovedForAll(
+            self: @ComponentState<TContractState>, owner: ContractAddress, operator: ContractAddress
+        ) -> bool {
+            self.get_approval_for_all(owner, operator).approved
+        }
+
+        fn setApprovalForAll(
+            ref self: ComponentState<TContractState>, operator: ContractAddress, approved: bool
+        ) {
+            self.set_approval_for_all(operator, approved)
+        }
+    }
+
+    ///
+    /// Internal
+    ///
+
+    #[generate_trait]
+    impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        +IWorldProvider<TContractState>,
+        impl ERC721Owner: erc721_owner_comp::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of InternalTrait<TContractState> {
+        // Helper function for allowance model
+        fn _get_approved(
```

Comment:
i like the _ prefix too xd, but we decided to follow cairo/rust convention so no _ prefix